cmd/codegen/codegenbuilder.go

```diff
@@ -0,0 +1,365 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"strings"
+	"text/template"
+
+	v1 "github.com/grafana/agent/pkg/integrations/v1"
+
+	"github.com/grafana/agent/pkg/integrations/shared"
+)
+
+type codeGen struct {
+}
+
+func (c *codeGen) generateConfigMeta() []configMeta {
+	configMetas := make([]configMeta, 0)
+	for _, i := range v1.Configs {
+		configMetas = append(configMetas, newConfigMeta(i.Config, i.DefaultConfig, i.Type))
+	}
+	return configMetas
+}
+
+func (c *codeGen) createV1Config() string {
+	configs := c.generateConfigMeta()
+	integrationTemplate, _ := template.New("shared").Parse(`
+
+type {{.Name}} struct {
+  {{.ConfigStruct}} ` + "`yaml:\",omitempty,inline\"`" + `
+  shared.Common ` + "`yaml:\",omitempty,inline\"`" + `
+}
+
+{{ if .DefaultConfig -}}
+func (c *{{ .Name }}) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	c.Config = {{ .PackageName }}.DefaultConfig
+	type plain {{ .Name }}
+	return unmarshal((*plain)(c))
+}
+{{ end -}}
+
+`)
+	v1Template, err := template.New("v1").Parse(`
+type V1Integration struct {
+  {{ range $index, $element := . -}}
+	{{ $element.Name }} *{{$element.Name}} ` + "`yaml:\"{{ $element.PackageName }},omitempty\"`\n" +
+		`{{ end -}}
+   TestConfigs []shared.V1IntegrationConfig ` + "`yaml:\"-,omitempty\"`\n" + `
+}
+
+func (v *V1Integration) ActiveConfigs() []shared.V1IntegrationConfig {
+    activeConfigs := make([]shared.V1IntegrationConfig,0)
+	{{ range $index, $element := . -}}
+	if v.{{ $element.Name }} != nil {
+        activeConfigs = append(activeConfigs, newConfigWrapper(&v.{{ $element.Name}}.Config, v.{{ $element.Name}}.Common))
+    }
+	{{ end -}}
+    for _, i := range v.TestConfigs {
+        activeConfigs = append(activeConfigs, i)
+    }
+    return activeConfigs
+}
```

Comment:
Serious question: do you feel like this is easier to understand than reflection?

Reply:
I don't mean to sound snarky, but genuinely want your opinion on if you think this is better than what we had before. I personally find complex text/templates pretty hard to follow and reason about, but I don't know how this compares

Reply:
Ultimately I think this will come down to what we feel is easier to understand, reason about, and maintain long term as integrations get more complex and we start supporting weirder niche use cases as the list grows.